File: codegen/src/main/java/net/webpdf/codegen/names/TypeName.java
package net.webpdf.codegen.names;

import java.nio.file.Path;

@SuppressWarnings("unused")
public class TypeName {

    private final String modelName;
    private final String pack;
    private final String name;

    public TypeName(String modelName) {
        String name = modelName;
        StringBuilder packageName = new StringBuilder();
        if (modelName.contains(".")) {
            String[] path = modelName.split("\\.");
            for (int i = 0; i < path.length; i++) {
                if (i != path.length - 1) {
                    if (i != 0) {
                        packageName.append(".");
                    }
                    packageName.append(path[i]);
                } else {
                    name = path[i];
                }
            }
        }
        this.modelName = modelName;
        this.pack = packageName.toString();
        this.name = name;
    }

    public TypeName(String pack, String name) {
        this.pack = pack;
        this.name = name;
        this.modelName = getPackageLocation("");
    }

    public TypeName(String modelName, String pack, String name) {
        this.modelName = modelName;
        this.pack = pack;
        this.name = name;
    }

    public String getModelName() {
        return modelName;
    }

    public String getPack() {
        return pack;
    }

    public String getName() {
        return name;
    }

    public String getPackageLocation(String basePath) {
        String path = getPackagePath(basePath);
        return path.isEmpty() ? name : path + "." + name;
    }

    public String getPackagePath(String basePath) {
        return basePath.isEmpty() ? pack :
                pack.isEmpty() ? basePath :
                        basePath + "." + pack;
    }

    public String getRootFileLocation() {
        return "./" + Path.of(pack.replaceAll("\\.", "/"))
                .toString().replaceAll("\\\\", "/") + "/" + name;
    }

    public String getRelativeFileLocation(String basePath) {
        String path = getRelativeFilePath(basePath);
        return path.isEmpty() ? name : path + "/" + name;
    }

    public String getRelativeFilePath(String basePath) {
        if (pack.equals(basePath)) {
            return ".";
        }
        Path current = Path.of(pack.replaceAll("\\.", "/"));
        Path target = Path.of(basePath.replaceAll("\\.", "/"));
        return "./" + target.relativize(current).toString().replaceAll("\\\\", "/");
    }

}

File: src/codegen/typescript/port/TypeName.ts
import path from "node:path";

/**
 * Utility wrapper for a fully qualified model name and path derivations.
 */
export class TypeName {
	private readonly modelName: string;
	private readonly pack: string;
	private readonly name: string;

	/**
	 * Creates a TypeName from a qualified model name (`pkg.Name`) or explicit parts.
	 */
	public constructor(modelName: string);
	public constructor(pack: string, name: string);
	public constructor(modelName: string, pack?: string, name?: string) {
		if (typeof pack === "string" && typeof name === "string") {
			this.modelName = modelName;
			this.pack = pack;
			this.name = name;
			return;
		}

		let resolvedName: string = modelName;
		let resolvedPackage: string = "";
		if (modelName.includes(".")) {
			const split: string[] = modelName.split(".");
			resolvedName = split[split.length - 1] ?? modelName;
			resolvedPackage = split.slice(0, split.length - 1).join(".");
		}

		this.modelName = modelName;
		this.pack = resolvedPackage;
		this.name = resolvedName;
	}

	/**
	 * Returns the full model name as provided.
	 */
	public getModelName(): string {
		return this.modelName;
	}

	/**
	 * Returns the package segment of the model.
	 */
	public getPack(): string {
		return this.pack;
	}

	/**
	 * Returns the short model name without package.
	 */
	public getName(): string {
		return this.name;
	}

	/**
	 * Returns the fully qualified package location relative to a base package.
	 *
	 * @param basePath Base package path.
	 */
	public getPackageLocation(basePath: string): string {
		const packagePath: string = this.getPackagePath(basePath);
		return packagePath.length === 0 ? this.name : `${packagePath}.${this.name}`;
	}

	/**
	 * Returns package path only (without model name), relative to a base package.
	 *
	 * @param basePath Base package path.
	 */
	public getPackagePath(basePath: string): string {
		if (basePath.length === 0) {
			return this.pack;
		}
		if (this.pack.length === 0) {
			return basePath;
		}
		return `${basePath}.${this.pack}`;
	}

	/**
	 * Returns root import location for the model.
	 */
	public getRootFileLocation(): string {
		return `./${path.join(this.pack.replace(/\./g, "/"), this.name).replace(/\\/g, "/")}`;
	}

	/**
	 * Returns the relative file location including model name.
	 *
	 * @param basePath Base package path.
	 */
	public getRelativeFileLocation(basePath: string): string {
		const relativePath: string = this.getRelativeFilePath(basePath);
		return relativePath.length === 0 ? this.name : `${relativePath}/${this.name}`;
	}

	/**
	 * Returns relative folder path between base package and model package.
	 *
	 * @param basePath Base package path.
	 */
	public getRelativeFilePath(basePath: string): string {
		if (this.pack === basePath) {
			return ".";
		}
		const current: string = this.pack.replace(/\./g, "/");
		const target: string = basePath.replace(/\./g, "/");
		return `./${path.relative(target, current).replace(/\\/g, "/")}`;
	}
}
